src/ray/core_worker/transport/actor_scheduling_queue.cc

```diff
@@ -107,22 +109,38 @@ void ActorSchedulingQueue::Add(
         task_spec,
         rpc::TaskStatus::PENDING_ACTOR_TASK_ARGS_FETCH,
         /* include_task_info */ false));
-    waiter_.Wait(dependencies, [seq_no, this]() {
-      RAY_CHECK(boost::this_thread::get_id() == main_thread_id_);
-      auto it = pending_actor_tasks_.find(seq_no);
-      if (it != pending_actor_tasks_.end()) {
-        const TaskSpecification &task_spec = it->second.TaskSpec();
-        RAY_UNUSED(task_event_buffer_.RecordTaskStatusEventIfNeeded(
-            task_spec.TaskId(),
-            task_spec.JobId(),
-            task_spec.AttemptNumber(),
-            task_spec,
-            rpc::TaskStatus::PENDING_ACTOR_TASK_ORDERING_OR_CONCURRENCY,
-            /* include_task_info */ false));
-        it->second.MarkDependenciesSatisfied();
-        ScheduleRequests();
+
+    auto in_actor_dependencies_metadata = task_spec.GetInActorDependenciesMetadata();
+    std::vector<rpc::ObjectReference> plasma_dependencies;
+    for (const auto &dep : dependencies) {
+      const auto &obj_id = ObjectID::FromBinary(dep.object_id());
+      // All object refs that don't have in-actor metadata are plasma refs.
+      if (!in_actor_dependencies_metadata.count(obj_id)) {
+        plasma_dependencies.push_back(dep);
       }
-    });
+    }
+
+    p2p_waiter_.Wait(in_actor_dependencies_metadata, [seq_no, task_spec, this]() {
```

Comment:
Step 5: Fetch the in-actor dependencies first, and then fetch in-plasma dependencies after in-actor deps finish. When both kinds of deps are ready, call `ScheduleRequests` to submit the tasks.

Reply:
The order may be not matter.